stdlib/public/Concurrency/Actor.cpp

```diff
@@ -0,0 +1,931 @@
+    return job->Flags.getKind() == JobKind::DefaultActor;
+  }
+};
+
+/// A job to process the default actor with a new priority; allocated
+/// separately from the actor.
+class ProcessOverrideJob;
+
+class JobRef {
+  enum : uintptr_t {
+    NeedsPreprocessing = 0x1,
+    IsOverride = 0x2,
+    JobMask = ~uintptr_t(NeedsPreprocessing | IsOverride)
+  };
+
+  /// A Job* that may have one of the two bits above mangled into it.
+  uintptr_t Value;
+
+  JobRef(Job *job, unsigned flags)
+    : Value(reinterpret_cast<uintptr_t>(job) | flags) {}
+public:
+  constexpr JobRef() : Value(0) {}
+
+  /// Return a reference to a job that's been properly preprocessed.
+  static JobRef getPreprocessed(Job *job) {
+    assert(job && "passing a null job");
+    return { job, 0 };
+  }
+
+  /// Return a reference to a job that hasn't been preprocesssed yet.
+  static JobRef getUnpreprocessed(Job *job) {
+    assert(job && "passing a null job");
+    return { job, NeedsPreprocessing };
+  }
+
+  /// Return a reference to an override job, which needs special
+  /// treatment during preprocessing.
+  static JobRef getOverride(ProcessOverrideJob *job);
+
+  /// Is this a null reference?
+  operator bool() const { return Value != 0; }
+
+  /// Does this job need to be pre-processed before we can treat
+  /// the job queue as a proper queue?
+  bool needsPreprocessing() const {
+    return Value & NeedsPreprocessing;
+  }
+
+  /// Is this an unpreprocessed override job?
+  bool isOverride() const {
+    return Value & IsOverride;
+  }
+
+  /// Given that this is an override job, return it.
+  ProcessOverrideJob *getAsOverride() const {
+    assert(isOverride());
+    return reinterpret_cast<ProcessOverrideJob*>(Value & JobMask);
+  }
+  ProcessOverrideJob *getAsPreprocessedOverride() const;
+
+  Job *getAsJob() const {
+    assert(!isOverride());
+    return reinterpret_cast<Job*>(Value & JobMask);
+  }
+  Job *getAsPreprocessedJob() const {
+    assert(!isOverride() && !needsPreprocessing());
+    return reinterpret_cast<Job*>(Value);
+  }
+
+  bool operator==(JobRef other) const {
+    return Value == other.Value;
+  }
+  bool operator!=(JobRef other) const {
+    return Value != other.Value;
+  }
+};
+
+/// The default actor implementation.
+class DefaultActorImpl : public HeapObject {
+  enum class Status {
+    /// The actor is not currently scheduled.  Implies that that the
+    /// job list is empty.
+    Idle,
+
+    /// There is currently a job scheduled to process the actor at the
+    /// stored max priority.
+    Scheduled,
+
+    /// There is currently a thread processing the actor at the stored
+    /// max priority.
+    Running
+  };
+
+  struct Flags : public FlagSet<size_t> {
+    enum : size_t {
+      Status_offset = 0,
+      Status_width = 3,
+
+      MaxPriority = 8,
+      MaxPriority_width = JobFlags::Priority_width,
+
+      // FIXME: add a reference to the running thread ID so that we
+      // can boost priorities.
```

Comment:
There's some redundancy here between the existence of jobs and the status, yeah.  And yeah, you can maybe use slightly different atomics if you have fixed buffer, although of course most of the expensive is in the barriers and contention rather than any specific complexity of the atomic instruction.